staging/kos/pkg/apis/network/types.go

```diff
@@ -177,9 +177,57 @@ type SubnetSpec struct {
 	VNI uint32
 }
 
+type SubnetConditionType string
+
+type ConditionStatus string
+
+const (
+	// ConditionTrue is the condition status which means a resource is in the
+	// condition.
+	ConditionTrue ConditionStatus = "True"
+
+	// ConditionFalse is the condition status which means a resource is not in
+	// the condition.
+	ConditionFalse ConditionStatus = "False"
+
+	// SubnetConflict is the condition that represents whether a subnet is in
+	// conflict with other subnets.
+	SubnetConflict SubnetConditionType = "Conflict"
+)
+
+type SubnetCondition struct {
+	// Type of the Subnet condition.
+	Type SubnetConditionType
+
+	// Status of the condition, one of True, False. In the future it might be
+	// extended with additional values, e.g. "Unknown".
+	Status ConditionStatus
+
+	// The reason for the condition's last transition.
+	// +optional
+	Reason string
+
+	// A human readable message indicating details about the transition.
+	// +optional
+	Message string
+}
+
 type SubnetStatus struct {
+	// Usable tells consumers whether it is safe to use the subnet or not. Once
+	// it becomes true, it is guaranteed to stay true unless an update to the
+	// subnet VNI or CIDR takes place.
```

Comment:
This comment makes some very specific assertions about a field with a very generic name.  I would be more comfortable with less stringent conditions or a more specific name (e.g., `OverlapFree`).

I think we should be focused on a simple example, and not stress too much about generality or future-proofing.  I would be fine with a completely specific name and comment.

Reply:
> I would be more comfortable with less stringent conditions

Do you mean making the comments less specific? I'm totally ok with that, but I don't mind the fact that the field has a general name: it makes it very clear to the consumer that the subnet can be used only when that field is true. I prefer that over something more specific. What is your problem with that? That the name does not convey anything regarding when it is set to `true` or `false`? Also, using a more specific name might be misleading to the user. Say that you use `OverlapFree` (but whatever gives away more details about how the value of the field is set works for my argument). That would be a lie. You could have two overlapping subnets S1 and S2, and S1 might be usable because it was created before (possibly very shortly before) S2. The reason I like `Usable` is that in the choice on whether a subnet can be used there is some "randomness": it does not only depend on the other subnets but on timing as well. I think `Valid` and more specific names (unless we pick an overly long and descriptive one) convey a more static meaning which is misleading. 

> I think we should be focused on a simple example, and not stress too much about generality or future-proofing. I would be fine with a completely specific name and comment, and not worrying about a general set of "conditions".

Along that line, I propose yet another change regarding the conditions. The reason I introduced them was to convey some information to the user/operator as for why the subnet is not usable. This is done by the Message field of the condition. But a condition is allowed to have more fields (e.g. Reason). We don't need that additional information in our example. Also, a condition can be true, false or unknown, but in our example it is only true. Since we only care about the error message, I propose to post only that to Subnets statuses. We can use the slice of strings already there for the error messages from the IPAM: 
```
type SubnetStatus struct {

...

	// Errors are the complaints, if any, from the IPAM controller.
	// +optional
	// +patchStrategy=replace
	Errors []string `json:"errors,omitempty" protobuf:"bytes,3,opt,name=errors" patchStrategy:"replace"`
}
```
If we need to distinguish the error messages from the subnet validator from those from the IPAM we can use the same pattern used for the errors in `NetworkAttachmentStatus` and wrap two distinct slices of strings within a struct:
```
type NetworkAttachmentStatus struct {
	// +optional
	Errors NetworkAttachmentErrors `json:"errors,omitempty" protobuf:"bytes,1,opt,name=errors"`

...

}
```
and
```
type NetworkAttachmentErrors struct {
	// IPAM holds errors about the IP Address Management for this attachment.
	// +optional
	// +patchStrategy=replace
	IPAM []string `json:"ipam,omitempty" protobuf:"bytes,1,opt,name=ipam" patchStrategy:"replace"`

	// Host holds errors from the node where this attachment is placed.
	// +optional
	// +patchStrategy=replace
	Host []string `json:"host,omitempty" protobuf:"bytes,2,opt,name=host" patchStrategy:"replace"`
}
```

Reply:
My complaint about associating the comment "Once it becomes true, it is guaranteed to stay true unless an update to the subnet VNI or CIDR takes place" with a field named `Usable` is that the comment says the field means more than just "usable".  A reviewer looking at client code, and seeing that it assumes that `Usable` does not change unless and until the Spec changes in certain ways, might think he has found a bug.  My suggestion of "OverlapFree" was a short-hand for a longer description.  One often has to settle for a short-hand, because putting the full definition in the name makes for unwieldy names.  But it is good to have a name that is at least suggestive of the meaning.  How about "SpecsValidated" (I use the plural "Specs" because the condition validated involves multiple objects), or simply "Validated" (omitting "Specs" on the grounds that it is only Specs that could reasonably be involved in "validation")?  I suggest "Validated" rather than just "Valid" because it connotes a judgement having been made, which might be asymmetric, and having been made at a point in time, and being a committment to that judgement.

If we go with a very generic name "Usable" then the comment should not make more specific promises about when that bit is true.  But I do not like going in the direction of erasing this particular information.  The validation of the VNI/CIDR/namespace consistency constraint is not an implementation detail that should be hidden.  It is feedback to the clients about whether they have done something wrong, and clients must know that this feedback is an important aspect of the lifecycle of a Subnet.

Reply:
I agree with putting the explanation in the `Errors` field.  I see that currently the IPAM controller does not put anything in SubnetStatus.Errors.  But that controller arguably _should_ put something in that field when it can not allocate an address because no more are available; this is a potentially transient condition, and `Errors` is allowed to report on potentially transient conditions.  So the solution of splitting the Errors field by source is plausible.

Reply:
Your arguments for changing `Usable` are sound, I'll change it. As you wrote this is a user-facing API so it makes sense to have expressive names. But I cannot think of something both truly expressive (i.e. giving more details, along the lines of `OverlapFree` --- which I don't like because it misses a lot of information) and concise, hence I favor `Validated` out of the proposed options. As you said it connotes that a judgement has been made.

Ok so I will switch from errors to conditions and make the IPAM write SubnetStatus.Errors when there's a shortage of addresses.

Reply:
"switch from errors to conditions"?  I thought we just agreed to go with errors.

Reply:
What I meant is: "switch from conditions to errors".